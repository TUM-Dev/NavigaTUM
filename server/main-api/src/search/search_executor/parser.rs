use std::collections::HashSet;

use log::warn;
use logos::Logos;

use crate::search::search_executor::lexer::Token;

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Filter {
    parents: HashSet<String>,
    types: HashSet<String>,
    usages: HashSet<String>,
}
impl Filter {
    pub fn as_meilisearch_filters(&self) -> String {
        let mut filters = vec![];
        if !self.parents.is_empty() {
            let parents: Vec<&str> = self.parents.iter().map(String::as_str).collect();
            filters.push(format!(
                "((parent_keywords IN {parents:?}) OR (parent_building_names IN {parents:?}) OR (campus IN {parents:?}))"
            ));
        }
        if !self.types.is_empty() {
            let types: Vec<&str> = self.types.iter().map(String::as_str).collect();
            filters.push(format!("(type IN {types:?})"));
        }
        if !self.usages.is_empty() {
            let usages: Vec<&str> = self.usages.iter().map(String::as_str).collect();
            filters.push(format!("(usage IN {usages:?})"));
        }
        filters.join(" AND ")
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Sorting {
    location: HashSet<String>,
}

impl Sorting {
    pub fn as_meilisearch_sorting(&self) -> Vec<String> {
        self.location
            .iter()
            .map(|s| format!("_geoPoint({s}):asc"))
            .collect()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TextToken {
    Text(String),
    SplittableText((String, String)),
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct ParsedQuery {
    pub tokens: Vec<TextToken>,
    pub filters: Filter,
    pub sorting: Sorting,
}

impl ParsedQuery {
    pub fn relevant_enough_for_room_highligting(&self) -> bool {
        if self.tokens.len() == 1 {
            return true;
        }
        match self.tokens.first() {
            Some(first) => match first {
                TextToken::Text(t) => t.len() > 3,
                TextToken::SplittableText((t0, _)) => t0.len() > 3,
            },
            None => false,
        }
    }
}

impl From<&str> for ParsedQuery {
    fn from(query: &str) -> Self {
        let mut result = Self::default();
        for token in Token::lexer(query) {
            match token {
                Ok(Token::Text(s)) => {
                    result.tokens.push(TextToken::Text(s));
                }
                Ok(Token::SplittableText((s1, s2))) => {
                    result.tokens.push(TextToken::SplittableText((s1, s2)));
                }
                Ok(Token::ParentFilter(filter)) => {
                    result.filters.parents.insert(filter);
                }
                Ok(Token::UsageFilter(filter)) => {
                    result.filters.usages.insert(filter);
                }
                Ok(Token::TypeFilter(filter)) => {
                    result.filters.types.insert(filter);
                }
                Ok(Token::LocationSort(location)) => {
                    result.sorting.location.insert(location);
                }
                Err(e) => {
                    warn!("Error in query parsing: {e:?}");
                }
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn parent_filter() {
        for filter in ["in:", "@"] {
            for sep in ["", " "] {
                assert_eq!(
                    ParsedQuery::from(format!("{filter}{sep}foo").as_str()).filters,
                    Filter {
                        parents: HashSet::from(["foo".to_string()]),
                        ..Default::default()
                    }
                );
            }
        }
    }

    #[test]
    fn usage_filters() {
        for filter in ["usage:", "nutzung:", "="] {
            for sep in ["", " "] {
                assert_eq!(
                    ParsedQuery::from(format!("{filter}{sep}foo").as_str()).filters,
                    Filter {
                        usages: HashSet::from(["foo".to_string()]),
                        ..Default::default()
                    }
                );
            }
        }
    }

    #[test]
    fn type_filters() {
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("type:{sep}foo").as_str()).filters,
                Filter {
                    types: HashSet::from(["foo".to_string()]),
                    ..Default::default()
                }
            );
        }
    }

    #[test]
    fn location_sort() {
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("near:{sep}45.32,59.3").as_str()).sorting,
                Sorting {
                    location: HashSet::from(["45.32,59.3".to_string()]),
                    ..Default::default()
                }
            );
        }
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("near:{sep}45.3,59.00000003").as_str()).sorting,
                Sorting {
                    location: HashSet::from(["45.3,59.00000003".to_string()]),
                    ..Default::default()
                }
            );
        }
    }

    #[test]
    fn text_token() {
        assert_eq!(
            ParsedQuery::from("foo").tokens,
            vec![TextToken::Text("foo".to_string())]
        );
        assert_eq!(
            ParsedQuery::from("foo foo").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
        assert_eq!(
            ParsedQuery::from("foo hs1").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::SplittableText(("hs".to_string(), "1".to_string())),
            ]
        );
    }

    #[test]
    fn text_filter_mixed() {
        assert_eq!(
            ParsedQuery::from("foo in:abc bar @abc foo near:45.32,59.3").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("bar".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
        assert_eq!(
            ParsedQuery::from(
                "foo in:abc bar @abc =def usage:dd nutzung:gh type:fdh foo near:45.32,59.3"
            )
            .tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("bar".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
    }
}
