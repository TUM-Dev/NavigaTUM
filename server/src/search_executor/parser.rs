use logos::Logos;
use std::collections::HashSet;
use std::fmt::{Debug, Formatter};
use tracing::warn;

use super::lexer::Token;

#[derive(Clone, Default, PartialEq, Eq)]
pub struct Filter {
    parents: HashSet<String>,
    types: HashSet<String>,
    usages: HashSet<String>,
}
impl Filter {
    pub fn as_meilisearch_filters(&self) -> String {
        let mut filters = vec![];
        if !self.parents.is_empty() {
            let parents: Vec<&str> = self.parents.iter().map(String::as_str).collect();
            filters.push(format!(
                "((parent_keywords IN {parents:?}) OR (parent_building_names IN {parents:?}) OR (campus IN {parents:?}))"
            ));
        }
        if !self.types.is_empty() {
            let types: Vec<&str> = self.types.iter().map(String::as_str).collect();
            filters.push(format!("(type IN {types:?})"));
        }
        if !self.usages.is_empty() {
            let usages: Vec<&str> = self.usages.iter().map(String::as_str).collect();
            filters.push(format!("(usage IN {usages:?})"));
        }
        filters.join(" AND ")
    }
    pub fn is_empty(&self) -> bool {
        self.parents.is_empty() && self.types.is_empty() && self.usages.is_empty()
    }
}

impl Debug for Filter {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("Filter");
        if !self.parents.is_empty() {
            base.field("parents", &self.parents);
        }
        if !self.types.is_empty() {
            base.field("types", &self.parents);
        }
        if !self.usages.is_empty() {
            base.field("usages", &self.parents);
        }
        base.finish()
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Sorting {
    location: HashSet<String>,
}

impl Sorting {
    pub fn as_meilisearch_sorting(&self) -> Vec<String> {
        self.location
            .iter()
            .map(|s| format!("_geoPoint({s}):asc"))
            .collect()
    }
    pub fn is_empty(&self) -> bool {
        self.location.is_empty()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TextToken {
    Text(String),
    SplittableText((String, String)),
}

#[derive(Clone, Default, PartialEq, Eq)]
pub struct ParsedQuery {
    pub tokens: Vec<TextToken>,
    pub filters: Filter,
    pub sorting: Sorting,
}

impl Debug for ParsedQuery {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("ParsedQuery");
        base.field("tokens", &self.tokens);
        if !self.filters.is_empty() {
            base.field("from", &self.filters);
        }
        if !self.sorting.is_empty() {
            base.field("from", &self.sorting);
        }
        base.finish()
    }
}

impl ParsedQuery {
    pub fn relevant_enough_for_room_highligting(&self) -> bool {
        if self.tokens.len() == 1 {
            return true;
        }
        match self.tokens.first() {
            Some(first) => match first {
                TextToken::Text(t) => t.len() > 3,
                TextToken::SplittableText((t0, _)) => t0.len() > 3,
            },
            None => false,
        }
    }
}

impl From<&str> for ParsedQuery {
    fn from(query: &str) -> Self {
        let mut result = Self::default();
        for token in Token::lexer(query) {
            match token {
                Ok(Token::Text(s)) => {
                    result.tokens.push(TextToken::Text(s));
                }
                Ok(Token::SplittableText((s1, s2))) => {
                    result.tokens.push(TextToken::SplittableText((s1, s2)));
                }
                Ok(Token::ParentFilter(filter)) => {
                    result.filters.parents.insert(filter);
                }
                Ok(Token::UsageFilter(filter)) => {
                    result.filters.usages.insert(filter);
                }
                Ok(Token::TypeFilter(filter)) => {
                    result.filters.types.insert(filter);
                }
                Ok(Token::LocationSort(location)) => {
                    result.sorting.location.insert(location);
                }
                Err(_) => {
                    warn!("Error in query parsing");
                }
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn parent_filter() {
        for filter in ["in:", "@"] {
            for sep in ["", " "] {
                assert_eq!(
                    ParsedQuery::from(format!("{filter}{sep}foo").as_str()).filters,
                    Filter {
                        parents: HashSet::from(["foo".to_string()]),
                        ..Default::default()
                    }
                );
            }
        }
    }

    #[test]
    fn usage_filters() {
        for filter in ["usage:", "nutzung:", "="] {
            for sep in ["", " "] {
                assert_eq!(
                    ParsedQuery::from(format!("{filter}{sep}foo").as_str()).filters,
                    Filter {
                        usages: HashSet::from(["foo".to_string()]),
                        ..Default::default()
                    }
                );
            }
        }
    }

    #[test]
    fn type_filters() {
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("type:{sep}foo").as_str()).filters,
                Filter {
                    types: HashSet::from(["foo".to_string()]),
                    ..Default::default()
                }
            );
        }
    }

    #[test]
    fn location_sort() {
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("near:{sep}45.32,59.3").as_str()).sorting,
                Sorting {
                    location: HashSet::from(["45.32,59.3".to_string()]),
                }
            );
        }
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("near:{sep}45.3,59.00000003").as_str()).sorting,
                Sorting {
                    location: HashSet::from(["45.3,59.00000003".to_string()]),
                }
            );
        }
    }

    #[test]
    fn text_token() {
        assert_eq!(
            ParsedQuery::from("foo").tokens,
            vec![TextToken::Text("foo".to_string())]
        );
        assert_eq!(
            ParsedQuery::from("foo foo").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
        assert_eq!(
            ParsedQuery::from("foo hs1").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::SplittableText(("hs".to_string(), "1".to_string())),
            ]
        );
    }

    #[test]
    fn text_filter_mixed() {
        assert_eq!(
            ParsedQuery::from("foo in:abc bar @abc foo near:45.32,59.3").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("bar".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
        assert_eq!(
            ParsedQuery::from(
                "foo in:abc bar @abc =def usage:dd nutzung:gh type:fdh foo near:45.32,59.3"
            )
            .tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("bar".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
    }
}
