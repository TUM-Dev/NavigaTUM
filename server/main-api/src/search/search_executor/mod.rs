use super::SanitisedSearchQueryArgs;
use cached::proc_macro::cached;
use log::error;

mod lexer;
mod parser;
mod postprocess;
mod query;
use crate::search::search_executor::parser::ParsedQuery;
use serde::Serialize;

#[derive(Serialize, Debug, Clone)]
pub struct SearchResultsSection {
    facet: String,
    entries: Vec<ResultEntry>,
    #[serde(skip_serializing_if = "Option::is_none")]
    n_visible: Option<usize>,
    #[serde(rename = "estimatedTotalHits")]
    estimated_total_hits: usize,
}

#[derive(Serialize, Debug, Clone)]
struct ResultEntry {
    id: String,
    r#type: String,
    name: String,
    subtext: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    subtext_bold: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    parsed_id: Option<String>,
}

// size=2500 seems to be about 250Mi
#[cached(size = 2500)]
pub async fn do_geoentry_search(
    q: String,
    highlighting: (String, String),
    args: SanitisedSearchQueryArgs,
) -> Vec<SearchResultsSection> {
    let parsed_input = ParsedQuery::from(q.as_str());

    match query::GeoEntryQuery::from(&parsed_input, &args, &highlighting)
        .execute()
        .await
    {
        Ok(response) => postprocess::merge_search_results(
            &args,
            &parsed_input,
            response.results.get(0).unwrap(),
            response.results.get(1).unwrap(),
            response.results.get(2).unwrap(),
            highlighting,
        ),
        Err(e) => {
            // error should be serde_json::error
            error!("Error searching for results: {e:?}");
            vec![]
        }
    }
}
