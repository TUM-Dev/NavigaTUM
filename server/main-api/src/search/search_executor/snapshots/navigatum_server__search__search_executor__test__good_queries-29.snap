---
source: main-api/src/search/search_executor/mod.rs
description: ""
expression: actual
info: "'0026m@5510' should get '5510.EG.026M' in 1 results"
---
- facet: rooms
  entries:
    - id: 5510.EG.026M
      type: room
      name: "\u00195510\u0017.EG.\u0019026M\u0017 (School Office Finanzen)"
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: MW 0026M@5510
    - id: 5510.EG.006M
      type: room
      name: "\u00195510\u0017.EG.\u0019006M\u0017 (WC-Herren)"
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: 0006M@5510
    - id: 5510.EG.026A
      type: room
      name: "\u00195510\u0017.EG.026A (Leitung Study & Teaching, Studienfachberatung)"
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: 0026A@5510
    - id: 5510.EG.026C
      type: room
      name: "\u00195510\u0017.EG.026C (IKOM-Besprechungsraum)"
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: 0026C@5510
    - id: 5510.EG.026B
      type: room
      name: "\u00195510\u0017.EG.026B (Studienbüro)"
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: 0026B@5510
    - id: 5510.EG.026N
      type: room
      name: "\u00195510\u0017.EG.026N (Study and Teaching Qualitätsmanagement)"
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: MW 0026N@5510
    - id: 5510.EG.029M
      type: room
      name: "\u00195510\u0017.EG.029M (Werkstatt)"
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: 0029M@5510
    - id: 5510.EG.028M
      type: room
      name: "\u00195510\u0017.EG.028M (Flur)"
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: 0028M@5510
    - id: 5531.EG.026
      type: room
      name: 5531.EG.026 (Beh.-WC / Personal)
      subtext: "garching, Ingeborg Ortner-Kinderhaus"
      subtext_bold: 0026@5531
    - id: 5501.EG.126M
      type: room
      name: 5501.EG.126M (Elektroverteiler)
      subtext: "garching, Maschinenwesen (MW)"
      subtext_bold: 0126M@5501
  n_visible: 10
  estimatedTotalHits: 15
- facet: sites_buildings
  entries: []
  n_visible: 0
  estimatedTotalHits: 0
